[liste des retours http](https://http.cat/)

les retours http sont une convention pour les api web et permette de savoir si ça c'est bien passé ou non grâce a un code.

1xx - Information (pour la machine)
2xx - Requête réussis
3xx - Redirection (pour la machine)
4xx - Erreur du client
5xx - Erreur du serveur

```python
from fastapi import FastAPI

app = FastAPI()

@app.post("/items/", status_code=201)
async def create_item(name: str):
    return {"name": name}
```

Grâce à FastAPI vous n'avez pas besoin de vous en souvenir, car il propose un module `status` qui est un Enum qui répertorie tous les types de status possibles.

```python
from fastapi import FastAPI, status

app = FastAPI()

@app.post("/items/", status_code=status.HTTP_201_CREATED)
async def create_item(name: str):
    return {"name": name}
```

<br>

### Exceptions

Vous pouvez lever les erreurs par vous même plutôt que de laisser la requête planter, grâce à `HTTPException` de fastapi.

```python
from fastapi import FastAPI, HTTPException

app = FastAPI()
items = {"foo": "The Foo"}

@app.get("/items/{item_id}")
async def read_item(item_id: str):
    if item_id not in items:
        raise HTTPException(status_code=404, detail="Item not found")
    return {"item": items[item_id]}
```

Sinon, on peut décider de créer sa propre exception HTTP et de la renvoyer au client.

```python
from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse

class UnicornException(Exception):
    def __init__(self, name: str):
        self.name = name

app = FastAPI()

@app.exception_handler(UnicornException)
async def unicorn_exception_handler(request: Request, exc: UnicornException):
    return JSONResponse(
        status_code=418,
        content={"message": f"Oops! {exc.name} did something. There goes a rainbow..."},
    )

@app.get("/unicorns/{name}")
async def read_unicorn(name: str):
    if name == "yolo":
        raise UnicornException(name=name)
    return {"unicorn_name": name}

```

<br>

